subt/control_center_qt.py

```diff
@@ -0,0 +1,283 @@
+                              QMessageBox, QMainWindow, QAction, QToolBar, QMenuBar, QCheckBox,
+                              QDockWidget)
+
+import osgar.record
+import osgar.logger
+import osgar.bus
+import osgar.node
+
+
+####################################################################################
+CFG_LORA = {
+  "version": 2,
+  "robot": {
+    "modules": {
+      "cc": {
+          "driver": "subt.control_center_qt:OsgarControlCenter",
+          "init": {}
+      },
+      "lora": {
+          "driver": "lora",
+          "init": {}
+      },
+      "lora_serial": {
+          "driver": "serial",
+          "init": {"port": "/dev/lora" if platform.system() == 'Linux' else 'COM35',
+                   "speed": 115200}
+      },
+    },
+    "links": [["lora_serial.raw", "lora.raw"],
+              ["lora.raw", "lora_serial.raw"],
+              ["lora.robot_status", "cc.robot_status"],
+              ["cc.cmd", "lora.cmd"]]
+  }
+}
+
+
+####################################################################################
+class DummyRobotReporter(osgar.node.Node):
+    def __init__(self, cfg, bus):
+        super().__init__(cfg, bus)
+        bus.register('status')
+        self.sleep_time = cfg['sleep']
+
+    def run(self):
+        step = 500
+        x, y = 0, 0
+        heading = 30
+        try:
+            while self.is_alive():
+                x += step * math.cos(math.radians(heading))
+                y += step * math.sin(math.radians(heading))
+                self.publish('status', [1, [int(x), int(y), heading*100], 'running'])
+                self.sleep(self.sleep_time)
+
+        except osgar.bus.BusShutdownException:
+            pass
+
+
+CFG_DEMO = {
+  "version": 2,
+  "robot": {
+    "modules": {
+      "cc": {
+          "driver": "subt.control_center_qt:OsgarControlCenter",
+          "init": {}
+      },
+      "robot1": {
+          "driver": "subt.control_center_qt:DummyRobotReporter",
+          "init": {
+              "sleep": 0.5
+          }
+      }
+    },
+    "links": [["robot1.status", "cc.robot_status"]]
+  }
+}
+####################################################################################
+
+
+class OsgarControlCenter:
+
+    def __init__(self, init, bus):
+        self.bus = bus
+        bus.register('cmd')
+        self.view = None # will be set from outside by record
+        self.thread = None
+
+    def start(self):
+        self.thread = threading.Thread(target=self.run)
+        self.thread.start()
+
+    def request_stop(self):
+        self.bus.shutdown()
+
+    def join(self):
+        self.thread.join()
+
+    def run(self):
+        while True:
+            try:
+                dt, channel, data = self.bus.listen()
+            except osgar.bus.BusShutdownException:
+                return
+            if channel == "robot_status":
+                robot, pose2d, status = data
+                self.view.robot_status.emit(robot, pose2d, status)
```

Comment:
probably here is should convert "integer pose2d" to meters and radians?? (can be postponed to next revision)

Reply:
the real question is, why we are sending the position in mm and 100th of degree in the first place. The position is computed in meters and radians anyway and serialization to the log file is lossless. I cannot remember the argument for this decision.